File: bound/golang/crypto/dsa/ed25519.go
package dsa

import (
	"errors"
	"web5/crypto"
	"web5/web5c"
)

type Ed25519Signer struct {
	cSigner *web5c.CEd25519Signer
}

func NewEd25519Signer(jwk crypto.JWK) (*Ed25519Signer, error) {
	cJWK := jwk.ToCJWK()

	signer, err := web5c.NewCEd25519Signer(cJWK)
	if err != nil {
		return nil, errors.New("failed to create Ed25519Signer")
	}

	return &Ed25519Signer{signer}, nil
}

func (s *Ed25519Signer) Sign(payload []byte) ([]byte, error) {
	return s.cSigner.Sign(payload)
}

File: bound/typescript/src/crypto/dsa/ed25519.ts
import wasm from "../../wasm";
import { Jwk } from "../jwk";
import { catchWeb5Error } from "../../errors";
import { Signer, TypescriptSigner } from ".";

export class Ed25519Generator {
  static generate(): Jwk {
    try {
      const wasmJwk = wasm.generate_ed25519_key(); 
      return Jwk.fromWasmJwk(wasmJwk); 
    } catch (error) {
      throw catchWeb5Error(error);
    }
  }
}

export class Ed25519Signer implements Signer {
  private signer: TypescriptSigner;

  constructor(jwk: Jwk) {
    try {
      this.signer = new TypescriptSigner(wasm.new_ed25519_signer(jwk.toWasmJwk()));
    } catch (error) {
      throw catchWeb5Error(error);
    }
  }

  sign(payload: Uint8Array): Uint8Array {
    return this.signer.sign(payload);
  }
}